services/offline/src/lib/clear-sensitive-caches.ts

```diff
@@ -0,0 +1,79 @@
+// IndexedDB names; should be the same as in @dhis2/pwa
+export const SECTIONS_DB = 'sections-db'
+export const SECTIONS_STORE = 'sections-store'
+
+// Non-sensitive caches that can be kept:
+const KEEPABLE_CACHES = [
+    /workbox-precache/, // precached static assets
+    /app-shell/, // will be immediately overwritten by network-first strategy
```

Comment:
@KaiVandivier is this not also sensitive?  It could include, for example, user information?  I'm assuming this is the requests *made by* the app-shell, such as `/system/info` and `/userSettings`

Reply:
Yes it is sensitive; my initial reasoning was that it shares the same downfall as other caches, i.e. that they can be inspected anyway, _but_ the I observed app shell cache can fail to be populated with all the necessary data to run offline if it gets cleared at the time when a new user is logging in. (i.e. 1. New user logs in, 2. In response, app adapter clears sensitive caches before app load, 3. App-shell cache gets incompletely populated due to staggered loading, 4. User goes offline, 5. App can't run offline because app-shell doesn't have all the necessary data)

So I figured that since the app-shell cache gets overwritten on page load, an app will never load with another user's app shell the same way it might with a cached section, so it won't have that problem, and the security aspect isn't much worse than the other caching behavior...

BUT now I have two questions to investigate:
1. Does the failed app shell also happen in production environments? It only be an issue for dev environment where nothing is precached
2. I think someone _could_ actually see someone else's app shell, if they log out as one user, go offline, and open up a PWA app... but maybe that's the same issue as not requiring authentication when offline

I'll take another look into the app shell failing and how precaching affects it though

Reply:
This implementation is also based off the assumption that we're more interested in avoiding the negative user experience where someone sees someone else's data than maintaining strict security

Reply:
Ah, found a better way: reloading a PWA app after clearing caches can safely restore its app-shell cache for the new user (and also solves some other issues): [adapter code](https://github.com/dhis2/app-platform/blob/e13a301dc0d8e1d5ec7cff8d2312ccf3d4daec27/adapter/src/utils/useVerifyLatestUser.js#L31-L38)

-- removed app-shell from keepable caches in a commit below

Reply:
👍  great!  Yeah, there are also some app-shell things like the apps menu that aren't loaded on startup, so they could be "stale" and leak data to new users visiting it when offline